### ISSUE TITLE
Presence of cats.implicits._ can cause type mismatches

### ISSUE THREAD

**Issue by paulcarey** [functional-streams-for-scala/fs2#1169]
This may not be a bug, but it is a rough edge and is likely to trip up newcomers who have imported `cats._, cats.implicits._` and expect them to play nicely with fs2.

```
scala> import cats.effect._
import cats.effect._

scala> import fs2._
import fs2._

scala> val s = Stream(1).flatMap(i => Stream.eval(IO { println(i) }))
s: fs2.Stream[cats.effect.IO,Unit] = Stream(..)

scala> import cats.implicits._
import cats.implicits._

scala> val s = Stream(1).flatMap(i => Stream.eval(IO { println(i) }))
<console>:20: error: type mismatch;
 found   : cats.effect.IO[Unit]
 required: fs2.Pure[?]
       val s = Stream(1).flatMap(i => Stream.eval(IO { println(i) }))
                                                     ^
```

**Comment by mpilquist** [comment]
Yeah, this is an unfortunate consequence of the trickery we play with `Pure` to make inference work nicely. Specifically, `Stream(1)` gives you a `Stream[Pure, Int]` -- when you have `cats.implicits._` imported, you get `flatMap` from `FlatMapSyntax`, which means you end up needing a function `[B] A => Stream[Pure, B]`.

Here's an example from the test suite: https://github.com/functional-streams-for-scala/fs2/blob/800ae03dbf34799ba98e898595e3c8dd1e43954a/core/shared/src/test/scala/fs2/CompilationTest.scala#L59-L64

You can work around it in a few ways -- either adding manual `covary` calls such that the cats provided `flatMap` ends up with a compatible signature, or importing `fs2.Stream.{PureOps, InvariantOps}` (one or both depending on what syntax you are using) *after* you've imported `cats.implicits._`.  E.g.:

```
import cats.implicits._
import fs2.Stream.{PureOps, InvariantOps}
val s = Stream(1).flatMap(i => Stream.eval(IO { println(i) }))
```

This would be a good addition to the faq.md file (PR appreciated!)

**Comment by smarter** [comment]
> Scala 3 might offer some relief via polymorphic function values + fixed inference bugs related to Nothing, the latter of which would enable us to replace Pure with Nothing and keep variance.

I'd love it If you could create some self-contained examples of type inference going wrong in Scala 2 and either add them as test cases to Dotty (it's as easy as dropping a file in the `tests/pos` directory) if they work currently or open an issue if they don't. We're trying to make inference as good as possible but it's hard to predict everything that people want to do with it.

**Comment by mpilquist** [comment]
@smarter Will do! I've tried some simple representative examples with Dotty and they've all passed. I won't go as far as to say there are no issues left with `Stream[Nothing, O]` but initial results were very promising. When I get around to writing up the overall design issue in more detail, I'll be sure to include Dotty in the writeup and add test cases if necessary.

**Comment by mpilquist** [comment]
@smarter Here's a good first example:

```scala
import scala.language.higherKinds

trait Stream[F[_], A]
object Stream {
  def emit[F[_], A](a: A): Stream[F, A] = new Stream[F, A] {}
  def eval[F[_], A](fa: F[A]): Stream[F, A] = new Stream[F, A] {}

  def append[F[_], A](x: Stream[F, A], y: Stream[F, A]): Stream[F, A] = new Stream[F, A] {}
}

trait IO[A]
object IO { def apply[A](f: => A): IO[A] = new IO[A]{} }

object Usage {
  val r = Stream.append(Stream.emit(1), Stream.eval(IO(1)))
}
```

scalac (2.12.6) fails with:
```
error: no type parameters for method append: (x: Stream[F,A], y: Stream[F,A])Stream[F,A] exist so that it can be applied to arguments (Stream[Nothing,Int], Stream[IO,Int])
 --- because ---
argument expression's type is not compatible with formal parameter type;
 found   : Stream[Nothing,Int]
 required: Stream[?F,?A]

  val r = Stream.append(Stream.emit(1), Stream.eval(IO(1)))
                 ^
01.scala:17: error: type mismatch;
 found   : Stream[Nothing,Int]
 required: Stream[F,A]
  val r = Stream.append(Stream.emit(1), Stream.eval(IO(1)))
                                   ^
01.scala:17: error: type mismatch;
 found   : Stream[IO,Int]
 required: Stream[F,A]
  val r = Stream.append(Stream.emit(1), Stream.eval(IO(1)))
                                                   ^
three errors found
```

dotc (0.7.0-RC1) works fine though.

**Comment by mpilquist** [comment]
Here's a longer writeup that talks about the general encoding we use for pure streams: https://mpilquist.github.io/blog/2018/07/04/fs2/

**Comment by paulcarey** [comment]
Thanks for the detailed explanation. Happy to update the FAQ. I'll submit a PR in a few days.

**Comment by mpilquist** [comment]
Fixed by #1173 in series/1.0
